File: rust/src/compat.rs
//! Backward-compatible aliases for the pre-1.14.0 "Pair" naming. 1.14.0
//! renamed Pair -> Copyset (types and methods) without shipping migration
//! aliases, a semver-breaking change in a minor release. This module
//! restores the old names as aliases/wrapper methods so code written
//! against <1.14.0 keeps compiling; new code should use the Copyset-named
//! exports directly. The old names are deprecated.
//!
//! `type X = Y` is a genuine Rust type alias (X and Y are the same type),
//! so this is a lossless, zero-cost compatibility layer for the type-only
//! renames below (`Pair`, `PairState`, `DrainPairStorageResponse`). It does
//! NOT cover field-level renames on structs that were never themselves
//! renamed (`BlockVolume::copyset_id`, `UpdateConfigResult`'s
//! `*copyset*`/`active_copyset_count_*` fields,
//! `VolumeBlockPlacementConfig::copyset_ids`,
//! `VolumeBlockPlacementResizeResult`'s `copyset_count_*`/`copysets_*`
//! fields): Rust has no field-alias mechanism, so restoring those would
//! need a hand-written parallel struct with its own Deserialize impl.
//! Neither is done here; no known Rust consumer of this SDK exists in the
//! mountOS suite. `UpdateVolumePairConfigRequest` is the one request type
//! below that had its own field renamed, so it is a distinct struct
//! translated at the call site rather than a plain alias.
//!
//! Hand-written, not touched by `make gen`.

use crate::{
    Copyset, CopysetState, DrainCopysetStorageResponse, Error, StoragesService,
    UpdateVolumeCopysetConfigRequest, VolumeBlockPlacementConfig, VolumeBlockPlacementResizeResult,
    VolumesService,
};

/// Deprecated: use [`CopysetState`] instead.
pub type PairState = CopysetState;

/// Deprecated: use [`Copyset`] instead.
pub type Pair = Copyset;

/// Deprecated: use [`DrainCopysetStorageResponse`] instead.
pub type DrainPairStorageResponse = DrainCopysetStorageResponse;

/// Deprecated: use [`UpdateVolumeCopysetConfigRequest`] instead.
#[derive(Debug, Clone)]
pub struct UpdateVolumePairConfigRequest {
    pub target_pair_count: i32,
}

impl From<&UpdateVolumePairConfigRequest> for UpdateVolumeCopysetConfigRequest {
    fn from(req: &UpdateVolumePairConfigRequest) -> Self {
        UpdateVolumeCopysetConfigRequest { target_copyset_count: req.target_pair_count }
    }
}

impl StoragesService {
    /// Deprecated: use [`StoragesService::list_copysets`] instead.
    pub async fn list_pairs(
        &self,
        storage_id: i64,
        state: Option<&str>,
        include_retired: Option<bool>,
    ) -> Result<Vec<Pair>, Error> {
        self.list_copysets(storage_id, state, include_retired).await
    }

    /// Deprecated: use [`StoragesService::get_copyset_status`] instead.
    pub async fn get_pair_status(&self, storage_id: i64, pair_id: &str) -> Result<Pair, Error> {
        self.get_copyset_status(storage_id, pair_id).await
    }

    /// Deprecated: use [`StoragesService::drain_copyset`] instead.
    pub async fn drain_pair(&self, storage_id: i64, pair_id: &str) -> Result<DrainPairStorageResponse, Error> {
        self.drain_copyset(storage_id, pair_id).await
    }
}

impl VolumesService {
    /// Deprecated: use [`VolumesService::get_copyset_config`] instead.
    pub async fn get_pair_config(&self, volume_id: i64) -> Result<VolumeBlockPlacementConfig, Error> {
        self.get_copyset_config(volume_id).await
    }

    /// Deprecated: use [`VolumesService::update_copyset_config`] instead.
    pub async fn update_pair_config(
        &self,
        volume_id: i64,
        req: &UpdateVolumePairConfigRequest,
    ) -> Result<VolumeBlockPlacementResizeResult, Error> {
        self.update_copyset_config(volume_id, &UpdateVolumeCopysetConfigRequest::from(req)).await
    }
}

File: go/compat.go
package sdk

import "context"

// Backward-compatible aliases for the pre-1.14.0 "Pair" naming. 1.14.0
// renamed Pair -> Copyset (types and methods) without shipping migration
// aliases, which is a semver-breaking change in a minor release. This file
// restores the old names as aliases/wrappers so code written against
// <1.14.0 keeps compiling. New code should use the Copyset-named exports
// directly; the old names are deprecated.
//
// Go type aliases (type X = Y) make Pair and Copyset the identical type, so
// this is a lossless, zero-cost compatibility layer for the type-only
// renames below (Pair, PairState, DrainPairStorageResponse). It does NOT
// cover field-level renames on struct types that were never themselves
// renamed (BlockVolume.PairID -> CopysetID, VolumeBlockPlacementConfig.PairIds,
// VolumeBlockPlacementResizeResult's PairCount*/Pairs* fields): Go has no
// field-alias mechanism, so restoring those would require either editing
// the generated structs (adding the old field back) or a hand-written
// parallel type with custom JSON decoding. Neither is done here; no known
// Go consumer of this SDK reads those fields (see mountos-launcher, its
// only in-repo Go consumer). UpdateVolumePairConfigRequest is the one
// request type below that had its own field renamed, so it is a distinct
// struct translated at the call site rather than a plain alias.
//
// Hand-written, not touched by `make gen`.

// Deprecated: use CopysetState instead.
type PairState = CopysetState

const (
	// Deprecated: use CopysetStateActive instead.
	PairStateActive = CopysetStateActive
	// Deprecated: use CopysetStateDraining instead.
	PairStateDraining = CopysetStateDraining
	// Deprecated: use CopysetStateSyncedDrained instead.
	PairStateSyncedDrained = CopysetStateSyncedDrained
	// Deprecated: use CopysetStateRetired instead.
	PairStateRetired = CopysetStateRetired
)

// Deprecated: use ParseCopysetState instead.
func ParsePairState(s string) (PairState, error) { return ParseCopysetState(s) }

// Deprecated: use Copyset instead.
type Pair = Copyset

// Deprecated: use DrainCopysetStorageResponse instead.
type DrainPairStorageResponse = DrainCopysetStorageResponse

// UpdateVolumePairConfigRequest is the pre-1.14.0 request shape for
// UpdatePairConfig. Its field was itself renamed (TargetPairCount ->
// TargetCopysetCount), so unlike the type-only renames above this cannot be
// a plain alias: it is a distinct type with the old field name, translated
// to UpdateVolumeCopysetConfigRequest by UpdatePairConfig below.
//
// Deprecated: use UpdateVolumeCopysetConfigRequest instead.
type UpdateVolumePairConfigRequest struct {
	TargetPairCount int32 `json:"targetPairCount"`
}

func (r *UpdateVolumePairConfigRequest) toCopysetConfigRequest() *UpdateVolumeCopysetConfigRequest {
	return &UpdateVolumeCopysetConfigRequest{TargetCopysetCount: r.TargetPairCount}
}

// Deprecated: use StoragesService.ListCopysets instead.
func (s *StoragesService) ListPairs(ctx context.Context, storageID int64, state string, includeRetired bool) ([]Pair, error) {
	return s.ListCopysets(ctx, storageID, state, includeRetired)
}

// Deprecated: use StoragesService.GetCopysetStatus instead.
func (s *StoragesService) GetPairStatus(ctx context.Context, storageID int64, pairID string) (*Pair, error) {
	return s.GetCopysetStatus(ctx, storageID, pairID)
}

// Deprecated: use StoragesService.DrainCopyset instead.
func (s *StoragesService) DrainPair(ctx context.Context, storageID int64, pairID string) (*DrainPairStorageResponse, error) {
	return s.DrainCopyset(ctx, storageID, pairID, nil)
}

// Deprecated: use VolumesService.GetCopysetConfig instead.
func (s *VolumesService) GetPairConfig(ctx context.Context, volumeID int64) (*VolumeBlockPlacementConfig, error) {
	return s.GetCopysetConfig(ctx, volumeID)
}

// Deprecated: use VolumesService.UpdateCopysetConfig instead.
func (s *VolumesService) UpdatePairConfig(ctx context.Context, volumeID int64, req *UpdateVolumePairConfigRequest) (*VolumeBlockPlacementResizeResult, error) {
	return s.UpdateCopysetConfig(ctx, volumeID, req.toCopysetConfigRequest())
}
